src/hotspot/share/opto/divnode.cpp

```diff
@@ -447,6 +447,81 @@ static Node *transform_long_divide( PhaseGVN *phase, Node *dividend, jlong divis
   return q;
 }
 
+template <typename TypeClass>
+Node* make_and(Node* a, Node* b);
+
+template <>
+Node* make_and<TypeLong>(Node* a, Node* b) {
+  return new AndLNode(a, b);
+}
+
+template <>
+Node* make_and<TypeInt>(Node* a, Node* b) {
+  return new AndINode(a, b);
+}
+
+template <typename TypeClass>
+Node* make_urshift(Node* a, Node* b);
+
+template <>
+Node* make_urshift<TypeLong>(Node* a, Node* b) {
+  return new URShiftLNode(a, b);
+}
+
+template <>
+Node* make_urshift<TypeInt>(Node* a, Node* b) {
+  return new URShiftINode(a, b);
+}
+
+template <typename TypeClass, typename Signed>
+Node* unsigned_div_ideal(PhaseGVN* phase, bool can_reshape, Node* div) {
+  // Check for dead control input
+  if (div->in(0) && div->remove_dead_region(phase, can_reshape)) {
+    return div;
+  }
+  // Don't bother trying to transform a dead node
+  if (div->in(0) && div->in(0)->is_top()) {
+    return nullptr;
+  }
+
+  const Type* t = phase->type(div->in(2));
+  if (t == TypeClass::ONE) { // Identity?
+    return nullptr;          // Skip it
+  }
+
+  const TypeClass* tl = t->is<TypeClass>();
+  if (!tl) {
+    return nullptr;
+  }
+
+  // Check for useless control input
+  // Check for excluding div-zero case
+  if (div->in(0) && (tl->_hi < 0 || tl->_lo > 0)) {
+    div->set_req(0, nullptr); // Yank control input
+    return div;
+  }
+
+  if (!tl->is_con()) {
+    return nullptr;
+  }
+  Signed l = tl->get_con(); // Get divisor
+
+  if (l == 0) {
+    return nullptr; // Dividing by zero constant does not idealize
+  }
+
+  if (l == min_jint) {
```

Comment:
Why excluding this case? You can move the check of division by 1 down here I think.

Reply:
You're right. That's an error. I was working on this just now, see my latest commit.